### PR TITLE
Remove ome- metadata handling from @loaders.gl/zarr WIP

### PR DESCRIPTION
**Pull request by manzt**
Suggested changes for #1441. Flattens `zarr` module, removes OME-specific metadata handling.

### CODE REVIEW

File: modules/zarr/src/lib/index.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+import { FetchFileStore } from './storage';
+import ZarrPixelSource from './pixel-source';
+import type { Labels } from '../types';
+
+/**
+ * Initializes multiscale Zarr pixel source.
+ * @param source Root url to multiscale Zarr store, or Zarr store.
+ * @param labels Labels for each array dimension.
+ */
+export async function load<S extends string[]>(source: string | ZarrArray['store'], labels: Labels<S>) {
```
Comment (manzt) [inline modules/zarr/src/lib/index.ts:13]: dimension `labels` are parsed from the OME metadata. We use labeled dimensions to unify indexing zarr/tiff sources, but something to consider is whether dimensions need labels at all.

Comment (manzt) [reply modules/zarr/src/lib/index.ts:13]: Ideally loaders.gl should work if just provided a URL (in either node or the browser). It would be great if a valid "store" could be substituted in the case where someone has a custom zarr-backend: https://github.com/manzt/reference-spec-reader/blob/f7b7487379add147763987496eb1a5917b43f2ef/src/store.js#L15

**Comment by kylebarron** [comment]
You and Ib might have discussed this, but I don't necessarily see an issue with _also_ having an OME Zarr loader. If it's possible to generalize the OME parts well, we might be able to have a core `ZarrLoader` as well as another `OMEZarrLoader` that extends the core functionality.

File: modules/zarr/src/index.ts [hunk 2]
```diff
@@ -1,3 +1 @@
-export { loadBioformatsZarr, loadOmeZarr } from './zarr';
-
-export { default as ZarrPixelSource } from './zarr/pixel-source';
+export { load } from './lib';
```
Comment (kylebarron) [inline modules/zarr/src/index.ts:1]: Nit: I'd suggest [having newlines](https://stackoverflow.com/q/729692) at the end of all files. IMO this is something loaders.gl CI (or probably ocular-dev-tools under the hood) should enforce.

File: modules/zarr/src/lib/index.ts (hunk 1 above)
Comment (kylebarron) [reply modules/zarr/src/lib/index.ts:13]: > We use labeled dimensions to unify indexing zarr/tiff sources

Is this similar to [xarray](https://github.com/pydata/xarray) labeling each dimension? It seems like a good fit for Zarr? I thought Zarr dimensions were always labeled...

File: modules/zarr/src/lib/storage.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
+import { fetchFile } from '@loaders.gl/core';
+import { KeyError } from 'zarr';
+import type { AsyncStore } from 'zarr/types/storage/types';
+
+/**
+ * Preserves (double) slashes earlier in the path, so this works better
+ * for URLs. From https://stackoverflow.com/a/46427607/4178400
```
Comment (kylebarron) [inline modules/zarr/src/lib/storage.ts:7]:
```suggestion
 * for URLs. From https://stackoverflow.com/a/46427607
```

FYI you don't need the last part of the SO url. It's just your user id, and IMO the shorter the URL the better for code comments.

**Comment by ibgreen** [comment]
> We might be able to have a core ZarrLoader as well as another OMEZarrLoader that extends the core functionality.

Yes. The overhead seems small, so why not keep it? 

Assuming the metadata is always in XML (unless that is an OME specific thing), it makes sense to parse that for the user regardless.

My main problem with the code is that the loadZarr function seems to call the loadOMEZarr function which seems upside-down.

If we can make it so that the loadOMEZarr calls loadZarr and does some extra processing that would be nicer. Perhaps loadOMEZarr just needs to supply a function that extracts some params from the metadata.

File: modules/zarr/src/lib/storage.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  }
+
+  async containsItem(key: string): Promise<boolean> {
+    const filepath = joinUrlParts(this.root, key);
+    const response = await fetchFile(filepath);
+    return response.ok;
+  }
+
+  keys(): Promise<string[]> {
+    return Promise.resolve([]);
```
Comment (kylebarron) [inline modules/zarr/src/lib/storage.ts:41]: Just curious is there a benefit to what you have over this?

```js
  async keys(): Promise<string[]> {
    return [];
```

File: modules/zarr/src/lib/storage.ts [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  }
+
+  setItem(): never {
+    throw new Error('setItem not implemented.');
```
Comment (kylebarron) [inline modules/zarr/src/lib/storage.ts:45]: Note I believe @ibgreen tends to take out error strings for bundle size reasons.

File: modules/zarr/src/lib/utils.ts [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+function prevPowerOf2(x: number) {
+  return 2 ** Math.floor(Math.log2(x));
+}
+
+export function guessTileSize(arr: ZarrArray) {
+  const interleaved = isInterleaved(arr.shape);
+  const [yChunk, xChunk] = arr.chunks.slice(interleaved ? -3 : -2);
+  const size = Math.min(yChunk, xChunk);
+  // deck.gl requirement for power-of-two tile size.
+  return prevPowerOf2(size);
```
Comment (kylebarron) [inline modules/zarr/src/lib/utils.ts:40]: How does this work when the inferred `tileSize` is no longer equal to the actual chunk size (when not already a power of two). Do we need any special handling for non-square chunks?

File: modules/zarr/src/lib/utils.ts [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+  };
+}
+
+function prevPowerOf2(x: number) {
+  return 2 ** Math.floor(Math.log2(x));
+}
+
+export function guessTileSize(arr: ZarrArray) {
+  const interleaved = isInterleaved(arr.shape);
+  const [yChunk, xChunk] = arr.chunks.slice(interleaved ? -3 : -2);
```
Comment (kylebarron) [inline modules/zarr/src/lib/utils.ts:37]: I think a comment might be helpful here to explain `-3` and `-2`.

Comment (kylebarron) [reply modules/zarr/src/lib/utils.ts:37]: Also I would've expected the chunk size to be part of the initial metadata in the `.zattrs` file?

File: modules/zarr/src/lib/utils.ts [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+      selection[dims(key as T)] = value as number;
+    }
+    return selection;
+  };
+}
+
+/*
+ * Helper method to determine whether pixel data is interleaved or not.
+ * > isInterleaved([1, 24, 24]) === false;
+ * > isInterleaved([1, 24, 24, 3]) === true;
```
Comment (kylebarron) [inline modules/zarr/src/lib/utils.ts:72]: Is the Zarr array shape always in a well-specified order like `x, y, z` (or I'm guessing `x, y, z` doesn't usually have a meaning in Zarr?)? 

Just wondering because with rasterio it's easy to get confused with band order because [it uses](https://rasterio.readthedocs.io/en/latest/topics/image_processing.html#imageorder) 

```
(bands, rows, columns)
```

while image processing software like `scikit-image`, `pillow` and `matplotlib` are generally ordered:

```
(rows, columns, bands)
```

Basically my question is: how do you know that a 3 or 4 in the last dimension means the data is interleaved?

File: modules/zarr/src/lib/storage.ts (hunk 4 above)
Comment (manzt) [reply modules/zarr/src/lib/storage.ts]: No benefit that I'm aware of. I just had an eslint warning.

File: modules/zarr/src/lib/index.ts (hunk 1 above)
Comment (manzt) [reply modules/zarr/src/lib/index.ts]: Somewhat similar. You can think of Zarr as large numpy arrays (with the option to add additional metadata). Without any extra metadata, Zarr arrays are "unlabeled". Libraries like `xarray` may use Zarr as a storage backend by adding structured metadata (JSON) in `.zattrs`. This metadata is separate from `.zarray` which contains the _array_-specific metadata (chunk size, shape, compression, dtype, etc).

File: modules/zarr/src/lib/utils.ts (hunk 6 above)
Comment (manzt) [reply modules/zarr/src/lib/utils.ts:79]: The ZarrPixelSource handles non-square chunking. It will fetch and stitch together additional chunks if necessary to satisfy the tilesize.

File: modules/zarr/src/lib/utils.ts (hunk 8 above)
Comment (manzt) [reply modules/zarr/src/lib/utils.ts]: A Zarr array is n-dimensional. Zarr.js only handles row-major ordering of arrays ("C" order), so we assume the last 2-3 dimensions are "special". If an image is interleaved, the last three dimension are `(rows, columns, bands)`, otherwise the last two dimensions are assumed to be `(rows, columns)`. The `rows` and `columns` dimensions are down-sampled across resolutions.

File: modules/zarr/src/lib/utils.ts (hunk 7 above)
Comment (manzt) [reply modules/zarr/src/lib/utils.ts:76]: `chunks` is included in the `.zarray` metadata (Array metadata). https://zarr.readthedocs.io/en/stable/spec/v2.html#arrays

This metadata initializes the `ZarrArray` object and can be accessed from `ZarrArray.chunks`. Zarr.js handles parsing `.zattrs`/`.zarray`.

**Comment by manzt** [comment]
> Assuming the metadata is always in XML (unless that is an OME specific thing), it makes sense to parse that for the user regardless.

It's not. Metadata for OME-Zarr is captured in the `.zattrs` (JSON). In Viv we also have support for a somewhat "hybird" intermediate format (we call "Bioformats Zarr") that is _not_ a standard. TL;DR the pixel data is the same as OME-Zarr (ZarrArrays) but the metadata is an OME-XML file that needs to be parsed separately. 

> My main problem with the code is that the loadZarr function seems to call the loadOMEZarr function which seems upside-down.

I have changed this now.

**Comment by manzt** [comment]
Not sure where to go from here with #1460

**Comment by ibgreen** [comment]
@manzt I landed the zarr loader in a separate PR (test cases commented out).

You will need to rebase/reapply these changes to master. 

Apologies for that, but it took a lot of iterations to get things sorted.
